### PR TITLE
Add CRI v1 proto

### PR DESCRIPTION
**Pull request by mrunalp**
@derekwaynecarr @SergeyKanzhelev @saschagrunert @mikebrow 

Signed-off-by: Mrunal Patel <mpatel@redhat.com>

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Add CRI v1beta proto and generated code.
We need both v1beta1 and v1alpha2 side by side so that
containerd and CRI-O could be updated. Once the runtimes
are updated and in the CI, we can switch the kubelet
to use v1beta1.

**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

- [KEP]: https://github.com/kubernetes/enhancements/issues/2040

```docs
None
```

### CODE REVIEW

File: hack/update-generated-runtime-dockerized.sh [hunk 1]
```diff
@@ -40,24 +42,34 @@ if [[ -z "$(which protoc)" || "$(protoc --version)" != "libprotoc 3."* ]]; then
...
 
-PATH="${gogopath}:${PATH}" \
-  protoc \
-  --proto_path="${KUBE_REMOTE_RUNTIME_ROOT}" \
-  --proto_path="${KUBE_ROOT}/vendor" \
-  --gogo_out=plugins=grpc:"${KUBE_REMOTE_RUNTIME_ROOT}" "${KUBE_REMOTE_RUNTIME_ROOT}/api.proto"
+function generate_code() {
+	RUNTIME_API_VERSION="$1"
+	KUBE_REMOTE_RUNTIME_PATH="${KUBE_REMOTE_RUNTIME_ROOT}/${RUNTIME_API_VERSION}"
+	PATH="${gogopath}:${PATH}" \
```
Comment (SergeyKanzhelev) [inline hack/update-generated-runtime-dockerized.sh:58]: not sure how it works in bash. PATH is a global variable. first call will set the first version. Second - will have the first version already set and add a new runtime to the end. Would it pick up the second runtime at all?

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/constants.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/constants.go:2]:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto [hunk 3]
```diff
@@ -0,0 +1,1338 @@
+/*
+Copyright 2018 The Kubernetes Authors.
```
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto:2]:
```suggestion
Copyright 2020 The Kubernetes Authors.
```

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto [hunk 4]
```diff
@@ -0,0 +1,1338 @@
...
+    // binary log files with structured logging data from the individual
+    // containers. For example, the files might be newline separated JSON
+    // structured logs, systemd-journald journal files, gRPC trace files, etc.
+    // E.g.,
+    //     PodSandboxConfig.LogDirectory = `/var/log/pods/<podUID>/`
+    //     ContainerConfig.LogPath = `containerName/Instance#.log`
+    //
+    // WARNING: Log management and how kubelet should interface with the
+    // container logs are under active discussion in
+    // https://issues.k8s.io/24677. There *may* be future change of direction
```
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto:367]: this issue is closed. Should we keep the warning around?

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto [hunk 5]
```diff
@@ -0,0 +1,1338 @@
...
+}
+
+message RunPodSandboxRequest {
+    // Configuration for creating a PodSandbox.
+    PodSandboxConfig config = 1;
+    // Named runtime configuration to use for this PodSandbox.
+    // If the runtime handler is unknown, this request should be rejected.  An
+    // empty string should select the default handler, equivalent to the
+    // behavior before this feature was added.
+    // See https://git.k8s.io/enhancements/keps/sig-node/runtime-class.md
```
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto:404]:
```suggestion
    // See https://git.k8s.io/enhancements/keps/sig-node/585-runtime-class/README.md
```

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto [hunk 6]
```diff
@@ -0,0 +1,1338 @@
...
+}
+
+message KeyValue {
+    string key = 1;
+    string value = 2;
+}
+
+// LinuxContainerResources specifies Linux specific configuration for
+// resources.
+// TODO: Consider using Resources from opencontainers/runtime-spec/specs-go
```
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto:567]: Are we keeping TODO in beta?

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto [hunk 7]
```diff
@@ -0,0 +1,1338 @@
...
+    //
+    // WARNING: Log management and how kubelet should interface with the
+    // container logs are under active discussion in
+    // https://issues.k8s.io/24677. There *may* be future change of direction
+    // for logging as the discussion carries on.
+    string log_path = 11;
+
+    // Variables for interactive containers, these have very specialized
+    // use-cases (e.g. debugging).
+    // TODO: Determine if we need to continue supporting these fields that are
```
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto:808]: should we keep this TODO in beta?

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto [hunk 8]
```diff
@@ -0,0 +1,1338 @@
...
+    rpc ListPodSandbox(ListPodSandboxRequest) returns (ListPodSandboxResponse) {}
+
+    // CreateContainer creates a new container in specified PodSandbox
+    rpc CreateContainer(CreateContainerRequest) returns (CreateContainerResponse) {}
+    // StartContainer starts the container.
+    rpc StartContainer(StartContainerRequest) returns (StartContainerResponse) {}
+    // StopContainer stops a running container with a grace period (i.e., timeout).
+    // This call is idempotent, and must not return an error if the container has
+    // already been stopped.
+    // TODO: what must the runtime do after the grace period is reached?
```
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto:69]: should we keep TODO in beta?

File: hack/update-generated-runtime-dockerized.sh (hunk 1 above)
Comment (mrunalp) [reply hack/update-generated-runtime-dockerized.sh:58]: Moving out of the function.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/constants.go (hunk 2 above)
Comment (mrunalp) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/constants.go]: Fixed

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto (hunk 3 above)
Comment (mrunalp) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto]: Fixed

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto (hunk 4 above)
Comment (mrunalp) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto]: Dropped the warning.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto (hunk 5 above)
Comment (mrunalp) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto]: Fixed, thanks!

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto (hunk 6 above)
Comment (mrunalp) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto]: Dropped the TODO.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto (hunk 7 above)
Comment (mrunalp) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto]: No, dropped.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto (hunk 8 above)
Comment (mrunalp) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1beta1/api.proto]: Replaced TODO by a line describing that the runtime should kill on grace period expiration.

**Comment by mikebrow** [comment]
thx for updating/removing the todo's I had a todo to do that and got swamped :-)

**Comment by cpuguy83** [comment]
This seems a little rushed to try and pull this off for v1.20. What's the benefit in doing it now vs waiting for v1.21?

**Comment by mrunalp** [comment]
> This seems a little rushed to try and pull this off for v1.20. What's the benefit in doing it now vs waiting for v1.21? 

We need a cycle to allow runtimes to support v1 before they can be brought back into CI and change the kubelet code. See - https://github.com/kubernetes/kubernetes/pull/96325#issuecomment-724899472

**Comment by cpuguy83** [comment]
Thanks @mrunalp, that makes sense then.

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Comment by dims** [comment]
/approve
/hold

@mrunalp please remove hold when you are ready

**Pull request merged**